libs/metall/src/dice/metall/detail/file.hpp

```diff
@@ -281,54 +279,165 @@ inline bool copy_file_dense(const std::filesystem::path &source_path,
+    METALL_ERRNO_ERROR("fallocate punch hole");
     return false;
   }
 
   return true;
 }
 
-inline bool copy_file_sparse_linux(const std::filesystem::path &source_path,
-                                   const std::filesystem::path &destination_path) {
-  try {
-    namespace bp = boost::process;
+/**
+ * Performs a sparse copy from src to dst, by only copying actual data and manually recreating
+ * all holes from src in dst.
+ *
+ * I.e. for each "segment" (data segment or hole) in src do
+ *      if segment is data => copy segment to dst
+ *      if segment is hole => create new hole (of the appropriate size) in dst
+ *
+ * @param src source file descriptor
+ * @param dst destination file descriptor
+ * @param src_size size of file behind src as obtained by ::fstat
+ * @return if copying was successful
+ *
+ * Relevant man pages:
+ *      https://www.man7.org/linux/man-pages/man2/lseek.2.html
+ *      https://www.man7.org/linux/man-pages/man2/copy_file_range.2.html
+ *      https://www.man7.org/linux/man-pages/man3/ftruncate.3p.html
+ */
+inline bool copy_file_sparse_linux(int src, int dst, off_t src_size) {
+  off_t old_off = 0;
+  off_t off = 0;
 
-    bp::ipstream is;
-    bp::child c{"/usr/bin/cp", "--sparse=always", source_path.c_str(), destination_path.c_str(),
-                bp::std_out > bp::null,
-                bp::std_err > is};
+  while ((off = ::lseek(src, off, SEEK_DATA)) >= 0) {
+    if (!create_hole_linux(dst, off - old_off)) {
+      METALL_ERROR("Unable to punch hole");
+      return false;
+    }
 
-    c.wait();
-    if (c.exit_code() == 0) {
-      fsync(destination_path);
-      return true;
+    off_t const hole_start = ::lseek(src, off, SEEK_HOLE);
+    if (hole_start < 0) {
+      METALL_ERRNO_ERROR("fseek(SEEK_HOLE)");
+      return false;
     }
 
-    std::string error;
-    std::copy(std::istreambuf_iterator<char>{is}, std::istreambuf_iterator<char>{}, std::back_inserter(error));
-    METALL_ERROR("Error during sparse copy: {}", error);
-  } catch (...) {
-    METALL_ERROR("Unable to spawn child /usr/bin/cp");
+    if (::copy_file_range(src, &off, dst, nullptr, hole_start - off, 0) < 0) {
+      METALL_ERRNO_ERROR("copy_file_range");
+      return false;
+    }
+
+    old_off = off;
+  }
+
+  if (errno != ENXIO) {
+    // error condition: offset is _not_ within a hole at the end of the file.
+    // previous lseek from while-loop condition must have failed
+    METALL_ERRNO_ERROR("fseek(SEEK_DATA)");
+    return false;
+  }
+
+  if (old_off < src_size) {
+    // the final extent is a hole we must call ftruncate
+    // here in order to record the proper length in the destination.
+    // See also: https://github.com/coreutils/coreutils/blob/a257b63ce7ebcc4577adb5406b39fc0edd61dcac/src/copy.c#L643-L658
+    if (::ftruncate(dst, src_size) < 0) {
+      METALL_ERRNO_ERROR("ftruncate");
+      return false;
+    }
+  }
+
+  if (!create_hole_linux(dst, src_size - old_off)) {
```

Comment:
This handles the case that source ends with an hole, right? Could this also be covered in `if (old_off < src_size) {` or is there another scenario where src_size - old_off is not 0.

Reply:
good point